frame/dapps-staking/src/pallet/mod.rs

```diff
@@ -795,18 +815,24 @@ pub mod pallet {
                     contract_id.clone(),
                     staker_reward,
                 ));
+
+                weight_info = T::WeightInfo::claim_staker_with_restake();
             }
 
-            T::Currency::resolve_creating(&staker, reward_imbalance);
+            // Withdraw reward funds from the dapps staking pot
+            let reward_imbalance = T::Currency::withdraw(
+                &Self::account_id(),
+                staker_reward,
+                WithdrawReasons::TRANSFER,
+                ExistenceRequirement::AllowDeath,
+            )?;
+
+            let reward_account = beneficiary.clone().unwrap_or(staker.clone());
+            T::Currency::resolve_creating(&reward_account, reward_imbalance);
             Self::update_staker_info(&staker, &contract_id, staker_info);
-            Self::deposit_event(Event::<T>::Reward(staker, contract_id, era, staker_reward));
+            Self::deposit_event(Event::<T>::Reward(reward_account, contract_id, era, staker_reward));
 
-            Ok(Some(if should_restake_reward {
-                T::WeightInfo::claim_staker_with_restake()
-            } else {
-                T::WeightInfo::claim_staker_without_restake()
-            })
-            .into())
+            Ok(Some(weight_info).into())
```

Comment:
Is this a personal preference or was the approach before incorrect?

Reply:
It's a personal preference, trying to make return logic simple and default based.